syntax = "proto3";
package pulse.study.bind.patient_variability;
option java_package = "com.kitware.pulse.study.bind.patient_variability";
option csharp_namespace = "pulse.study.bind.patient_variability";
option optimize_for = SPEED;

import "pulse/cdm/bind/Patient.proto";
import "pulse/cdm/bind/TestReport.proto";
import "pulse/engine/bind/Scenario.proto";

enum eGenStyle
{
  Combinatorial = 0; /**<<@brief Classic combinatorial of all factors. */
  Slice         = 1; /**<<@brief Iterate 1 property at a time, hold other properties to their slice index value. */
}

message ParameterIterationData
{
  uint32          SliceIndex = 1;
  repeated double Values     = 2;
}

message PatientIterationData
{
  eGenStyle                                        GenStyle                    = 1;
  pulse.cdm.bind.PatientData.eSex                  Sex                         = 2;
  ParameterIterationData                           Age_yr                      = 3;
  ParameterIterationData                           Height_cm                   = 4;
  ParameterIterationData                           BodyMassIndex               = 5;
  ParameterIterationData                           BodyFatFraction             = 6;
  ParameterIterationData                           HeartRate_bpm               = 7;
  ParameterIterationData                           MeanArtierialPressure_mmHg  = 8;
  ParameterIterationData                           PulsePressure_mmHg          = 9;
  ParameterIterationData                           RespirationRate_bpm         = 10;
}

message ScenarioIterationData
{
  pulse.engine.bind.ScenarioData                   Scenario                    = 1;
  eGenStyle                                        GenStyle                    = 2;
  bool                                             CreateStates                = 3;
  double                                           BaselineDuration_s          = 4;
  double                                           MaxSimTime_min              = 5;
}

message TCCCIterationData
{
  ScenarioIterationData                            ScenarioIteration           = 1;
  bool                                             PerformInterventions        = 2;// Set to false to study injury effects only
  ParameterIterationData                           AirwayObstructionSeverity   = 3;
  ParameterIterationData                           HemorrhageSeverity          = 4;
  ParameterIterationData                           TensionPneumothoraxSeverity = 5;
  ParameterIterationData                           InsultDuration_s            = 6;
  ParameterIterationData                           SalineAvailable             = 7;// Should only contain 0 and/or 1
  ParameterIterationData                           NeedleAvaiable              = 8;// Should only contain 0 and/or 1
  ParameterIterationData                           ChestWrapAvailable          = 9;// Should only contain 0 and/or 1
}

message PatientStateData
{
  /** @brief Enumeration for engine failure types */
  enum eFailure
  {
    None=0;               /**<< @brief No stabilization failures. */
    FailedSetup=1;        /**<< @brief SetupPatient could not create a valid Pulse patient with provided patient values. */
    FailedStabilization=2;/**<< @brief Pulse was unable to stabilize to provided/computed patient values. */
    RuntimeError=3;       /**<< @brief Error running the engine with (optional) actions. */
    FatalRuntimeError=4;  /**<< @brief Fatal error running the engine with (optional) actions. */
  }

  uint32                                                  ID                                = 1;
  string                                                  OutputBaseFilename                = 2;
  string                                                  StateFilename                     = 3;
  string                                                  Description                       = 4;
  double                                                  MaxSimulationTime_s               = 5;
  double                                                  FinalSimulationTime_s             = 6;/**<< @brief We could stop the sim early (ex. due to hemorrhage) */
  // Requested Patient
  pulse.cdm.bind.PatientData                              Patient                           = 7;
  pulse.cdm.bind.PatientData                              SetupPatient                      = 8;
  // Stabilization Info
  eFailure                                                Failure                           = 9;
  double                                                  StabilizationTime_s               = 10;

  message ValidationData
  {
    map<string,pulse.cdm.bind.PropertyValidationListData> ValidationMap                     = 1;
  }
}

message PatientStateListData
{
  repeated PatientStateData      PatientState               = 1;
}
